packages/contracts/src/contracts/current/protocol/Exchange/MixinSignatureValidator.sol

```diff
@@ -169,9 +170,38 @@ contract MixinSignatureValidator is
             isValid = signer == recovered;
             return isValid;
 
-        // Signature verified by signer contract
+        // Signature verified by signer contract.
+        // If used with an order, the maker of the order is the signer contract.
         } else if (signatureType == SignatureType.Contract) {
-            isValid = ISigner(signer).isValidSignature(hash, signature);
+            // Pass in signature without signature type.
+            bytes memory signatureWithoutType = deepCopyBytes(
+                signature,
+                1,
+                signature.length - 1
+            );
+            isValid = ISigner(signer).isValidSignature(hash, signatureWithoutType);
+            return isValid;
+
+        // Signature verified by validator contract.
+        // If used with an order, the maker of the order can still be an EOA.
+        // A signature using this type should be encoded as:
+        // | Offset | Length | Contents                        |
+        // | 0x00   | 1      | Signature type is always "\x07" |
+        // | 0x01   | 20     | Address of validator contract   |
+        // | 0x15   | **     | Signature to validate           |
+        } else if (signatureType == SignatureType.Validator) {
```

Comment:
Maybe I'm missing something obvious here, with the Signature Contract the makerAddress is the signer, so we call back into the contract of Signer (which is makerAddress) to validate the signature. This shows intent as both the makerAddress and the validator are the same.

In the Signature Validator case a separate Validator is called (not makerAddress). How do we know the makerAddress intended to use the Signature Validator? Couldn't I deploy a malicious Validator that accepts any signatures and create fake orders using that Validator?

Should there be an Authorized Validator check for signer?

Reply:
Yep, you're right. There needs to be a way of approving a validator contract.